### PR TITLE
Free function api

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
A free function API for cq.Shape. Originally inspired by #1269.

The idea is to provide a low-level but reasonably user friendly API for cq.Shape

- [x] primitives
- [x] operations
- [x] bool op as operators
- [x] moved with multiple locations
  - [x]  Overload for `*locs`
  - [x] overload for `moved(x=0, y=0, z=0, rx=0, ry=0, rz=0)` 
- [x] Location constructor improvements ?
- [x] additional factories ?
- [x] tests
- [x] docs

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
@lorenzncode will you have bandwidth/interest to look at this?

**Comment by lorenzncode** [comment]
@adam-urbanczyk Yes, at first glance this looks very interesting.  I'll find some time to take a look.

**Comment by lorenzncode** [comment]
I started experimenting.  It seems very easy to use!  Can revolve be changed to accept Shell?

```py
    # https://groups.google.com/g/cadquery/c/B-9w-U3gs1M/m/YhwbkMYPCAAJ
    from cadquery.occ_impl.shapes import spline, loft, revolve

    pts0 = [(1, 1, 0), (1, 1, 1), (1.2, 0.7483, 2)]
    pts1 = [(2, 2, 0), (2, 2, 1), (2.2, 1.7776, 2)]
    edge0 = spline(pts0)
    edge1 = spline(pts1)

    sh0 = loft([edge0, edge1])
    result = revolve(sh0, pts0[0], pts1[0], 180) # changed revolve _get Shell
```
[test_free_api_jupyter.zip](https://github.com/CadQuery/cadquery/files/14217478/test_free_api_jupyter.zip)

**Comment by adam-urbanczyk** [comment]
I added shape normalization. For your example `loft` returns `Face` instead of `Shell`.

**Comment by lorenzncode** [comment]
Very exciting new CQ development!  I checked out the branch and started trying it out.  Will play around with it more.

File: cadquery/occ_impl/shapes.py [hunk 1]
```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+
+        results.append(builder.Shape())
+
+    return _compound_or_shape(results)
+
+
+@sweep.register
+def sweep(s: Sequence[Shape], path: Shape, cap: bool = False) -> Shape:
+    """
+    Sweep edges or wires along a path, chaning sections are supported.
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:4916]:
```suggestion
    Sweep edges or wires along a path, chaining sections are supported.
```

File: doc/free-func.rst [hunk 2]
```diff
@@ -0,0 +1,236 @@
+.. _freefuncapi:
+
+*****************
+Free function API
+*****************
+
+.. warning:: The free function API is experimental and may change.
+
+For situations when more freedom in crating individual objects is required a free function API is provided.
```
Comment (jmwright) [inline doc/free-func.rst:9]:
```suggestion
For situations when more freedom in crafting individual objects is required, a free function API is provided.
```

File: doc/free-func.rst [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    top = cap(side.edges('>Z'), side, [(0,0,1.6*dh)])
+
+    # assemble into a solid
+    s = solid(side, bottom, top)
+
+    # construct the final result
+    result = s.moved((-3*r, 0, 0), (3*r, 0, 0))
+
+
+The code above builds non-trivial object by sequentially constructing individual faces, assembling them into a solid and finally generating a pattern.
```
Comment (jmwright) [inline doc/free-func.rst:50]:
```suggestion
The code above builds a non-trivial object by sequentially constructing individual faces, assembling them into a solid and finally generating a pattern.
```

File: doc/free-func.rst [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+It begins with defining few edges.
+
+.. code-block:: python
+
+    edge1 = circle(r)
+    edge2 = circle(2*r).moved(z=dh)
+    edge3 = circle(r).moved(z=1.5*dh)
+
+
+Those edges are used to crate the side faces of the final solid using loft.
```
Comment (jmwright) [inline doc/free-func.rst:61]: There are a few places like this (`loft`) that mention method names. Should they be turned into reference links to the API docs?

File: doc/free-func.rst [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    s = solid(side, bottom, top)
+
+Finally, the solid is duplicated and placed in the desired locations creating the final compound object. Note various usages of moved.
+
+.. code-block:: python
+
+    result = s.moved((-3*r, 0, 0), (3*r, 0, 0))
+
+In general all the operations are implemented as free functions, with the exception of placement and selection  which are strictly related to a specific shape.
```
Comment (jmwright) [inline doc/free-func.rst:90]:
```suggestion
In general all the operations are implemented as free functions, with the exception of placement and selection which are strictly related to a specific shape.
```

File: doc/free-func.rst [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    c = circle(1)
+
+    f = plane(1, 1.5)
+
+    b = box(1, 1, 1)
+
+    result = compound(e, c.move(2), f.move(4), b.move(6))
+
+
+Boolean operations
```
Comment (jmwright) [inline doc/free-func.rst:113]: Are there any performance issues here that should be mentioned?

File: doc/free-func.rst [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    r4 = compound(s1, s2, s3)
+
+    results = (r1, r2, r3, r4,)
+    result = compound([el.moved(2*i) for i,el in enumerate(results)])
+
+
+Operations
+----------
+
+Free function api currently supports extrude, loft, revolve and sweep operations.
```
Comment (jmwright) [inline doc/free-func.rst:193]:
```suggestion
Free function API currently supports extrude, loft, revolve and sweep operations.
```

File: mypy/cadquery-plugin.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    else:
+        return_type_names = [ctx.args[1][0].value]
+
+    return UnionType([ctx.api.named_type(n) for n in return_type_names])
+
+
+def hook__get_wires(ctx: FunctionContext) -> Type:
+    """
+    Hook for cq.occ_impl.shapes._get_wires
+
```
Comment (jmwright) [inline mypy/cadquery-plugin.py:62]:
```suggestion
```

File: mypy/cadquery-plugin.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    return_type = ctx.api.named_type("Wire")
+
+    return ctx.api.named_generic_type("typing.Iterable", [return_type])
+
+
+def hook__get_edges(ctx: FunctionContext) -> Type:
+    """
+    Hook for cq.occ_impl.shapes._get_edges
+
```
Comment (jmwright) [inline mypy/cadquery-plugin.py:73]:
```suggestion
```

**Review by jmwright — APPROVED** [review]

**Comment by jmwright** [comment]
Thanks for all the work on this @adam-urbanczyk ! The examples run fine for me, and the code looks good.

Either in this PR or a separate PR, we will also need to update setup.py for the new multimethod version declaration. With version 1.9.1 I get an intermittent error. This means we will probably see some support issues after the next release where people did not upgrade their multimethod version or create a fresh environment.

File: cadquery/occ_impl/shapes.py [hunk 10]
```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+def _shapes_to_toptools_list(s: Iterable[Shape]) -> TopTools_ListOfShape:
+    rv = TopTools_ListOfShape()
+
+    for el in s:
+        rv.Append(el.wrapped)
+
+    return rv
+
+
+# %% alternative constructors
```
[the remark this answers is not in the document]
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:4282]: See above, space is just a typo.

File: cadquery/occ_impl/shapes.py [hunk 11]
```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
     ShapeAnalysis_FreeBounds.ConnectEdgesToWires_s(edges_in, tol, False, wires_out)
 
     return [Wire(el) for el in wires_out]
+
+
+#%% utilities
```
[the remark this answers is not in the document]
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:4099]: Some IDEs (at least spyder) partition the functions/classes into sections based on `%%`.

File: cadquery/occ_impl/shapes.py [hunk 12]
```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+        builder.Build()
+
+        results.append(builder.Shape())
+
+    return _compound_or_shape(results)
+
+
+def offset(s: Shape, t: float, cap=True, tol: float = 1e-6) -> Shape:
+    """
+    Offset or thicken faces or shells.
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4863]: @adam-urbanczyk  Would you consider splitting the offset function into two separate functions `offset` and `thicken`?

The `cap` parameter could then be eliminated. If a thicken function is implemented, what do you think about providing a parameter similar to `Workplane.extrude` `both` to optionally thicken in both directions?

Say if thicken function was available with both=True:
```py
# thicken shell sh0 in both directions
s0 = thicken(sh0, 1, True)
```

It is still possible with current PR to thicken in both directions:
```py
# thicken shell sh0 in both directions
s0 = offset(sh0, 1)
s1 = offset(sh0, -1)
s2 = s0 + s1
```

The thicken function would be comparable to the thicken Face method except that it also supports Shell.
```py
e0 = segment((0, 0), (10, 10))
f0 = extrude(e0, (0, 0, 10))
s0 = f0.thicken(1)   # existing Face method
s1 = thicken(f0, 1)  # this PR
```

I'm not suggesting it is a Must fix/change if you are against it and already rejected the idea.

File: cadquery/occ_impl/shapes.py [hunk 13]
```diff
@@ -3984,3 +4094,875 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+    return Compound(rv)
+
+
+@compound.register
+def compound(s: Sequence[Shape]) -> Shape:
+
+    return compound(*s)
+
+
+#%% primitives
```
[the remark this answers is not in the document]
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:4396]: See above

File: cadquery/occ_impl/shapes.py (hunk 12 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:4870]: Currently `offset` has the same behavior as `Face.thicken`, i.e. it is not symmetric. I don't want to add more functionality to this PR, but the API will for sure change - it should be for now considered experimental. `thicken` free func could be added in future PRs.

File: doc/free-func.rst (hunk 4 above)
Comment (adam-urbanczyk) [reply doc/free-func.rst]: Done

File: doc/free-func.rst (hunk 6 above)
Comment (adam-urbanczyk) [reply doc/free-func.rst:113]: Done

**Comment by adam-urbanczyk** [comment]
OK, I think I resolved all the points. Let me know if you are OK with merging.

File: doc/free-func.rst [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+It begins with defining few edges.
+
+.. code-block:: python
+
+    edge1 = circle(r)
+    edge2 = circle(2*r).moved(z=dh)
+    edge3 = circle(r).moved(z=1.5*dh)
+
+
+Those edges are used to crate the side faces of the final solid using :meth:`~cadquery.occ_impl.shapes.loft`.
```
Comment (lorenzncode) [inline doc/free-func.rst:61]: typo; crate->create

File: doc/free-func.rst [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+
+    result = compound(e, c.move(2), f.move(4), b.move(6))
+
+
+Boolean operations
+------------------
+
+Boolean operations are supported and implemented as operators and free functions.
+In general boolean operations are slow and it is advised to avoid them and not to perform the in a loop.
+One can for example union multiple solids at once by irst combining them into a compound.
```
Comment (lorenzncode) [inline doc/free-func.rst:118]: typo: irst->first

File: cadquery/occ_impl/shapes.py [hunk 16]
```diff
@@ -1462,6 +1544,34 @@ def _siblings(shapes, level):
...
+    def __sub__(self, other: "Shape") -> "Shape":
+        """
+        Subtract other from self.
+        """
+
+        return cut(self, other)
+
+    def __mul__(self, other: "Shape") -> "Shape":
+        """
+        Intersect slef and other.
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:1563]: typo: slef->self

File: cadquery/occ_impl/shapes.py [hunk 17]
```diff
@@ -3984,3 +4094,882 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+                raise ValueError(
+                    f"Required type(s): {types}; encountered {el.ShapeType()}"
+                )
+    else:
+        raise ValueError(f"Required type(s): {types}; encountered {t}")
+
+
+def _get_one(s: Shape, ts: Union[Shapes, Tuple[Shapes, ...]]) -> Shape:
+    """
+    Get one shape or raise an error
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4132]: docstrings: For consistency add period here?
Get one shape or raise an error
->
Get one shape or raise an error.

Period missing in following four functions as well.

File: cadquery/occ_impl/shapes.py [hunk 18]
```diff
@@ -3984,3 +4094,882 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+
+    for i, p in enumerate(ps):
+        rv.SetValue(i + 1, Vector(p).toPnt())
+
+    return rv
+
+
+def _shapes_to_toptools_list(s: Iterable[Shape]) -> TopTools_ListOfShape:
+    """
+    Convert an itrable of Shape to a TopTools list (OCCT specific).
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4278]: typo: itrable->iterable

File: cadquery/occ_impl/shapes.py [hunk 19]
```diff
@@ -3984,3 +4094,882 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+@wire.register
+def wire(s: Sequence[Shape]) -> Shape:
+
+    return wire(*s)
+
+
+@multimethod
+def face(*s: Shape) -> Shape:
+    """
+    Builld face from edges or wires
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4315]: typo: Builld ->Build

Add period to face docstring?  Add period also to next functions shell, solid?

File: cadquery/occ_impl/shapes.py [hunk 20]
```diff
@@ -3984,3 +4094,882 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+@face.register
+def face(s: Sequence[Shape]) -> Shape:
+
+    return face(*s)
+
+
+@multimethod
+def shell(*s: Shape) -> Shape:
+    """
+    Build shell form faces
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4340]: typo: form->from

File: cadquery/occ_impl/shapes.py [hunk 21]
```diff
@@ -3984,3 +4094,882 @@ def edgesToWires(edges: Iterable[Edge], tol: float = 1e-6) -> List[Wire]:
...
+        builder.Add(Vector(p).toPnt())
+
+    builder.Close()
+
+    return _compound_or_shape(builder.Wire())
+
+
+def rect(w: float, h: float) -> Shape:
+    """
+    Construct a rectangle
```
Comment (lorenzncode) [inline cadquery/occ_impl/shapes.py:4461]: docstring period?

**Review by lorenzncode — CHANGES_REQUESTED** [review]
I've added few comments on docs/docstrings after proofreading.

**Comment by lorenzncode** [comment]
Thank you for the new API!

**Review by lorenzncode — APPROVED** [review]

**Comment by jmwright** [comment]
+1 to merge

**Pull request merged**